### ISSUE TITLE
Adversarial Loss Implementation Mismatch from StreamVC

### ISSUE THREAD

**Issue by arishov1** [yuval-reshef/StreamVC#2]
Hi,

Thank you for your implementation.

As I was reading the StreamVC paper, I noticed that they use the adversarial loss as described [here](https://arxiv.org/pdf/2107.03312). If we implement the loss as described in that paper, the code will change as follows:

Before:
```
class GeneratorLoss(nn.Module):
    def forward(self, fake: list[list[torch.Tensor]], mask_ratio: torch.Tensor):
        loss = torch.tensor(
            0., device=fake[0][0].device, dtype=fake[0][0].dtype)
        for scale in fake:
            loss += -masked_mean_from_ratios(scale[-1], mask_ratio)
        return loss
```
        
 After:
```
 class GeneratorLoss(nn.Module):
    def forward(self, fake: list[list[torch.Tensor]], mask_ratio: torch.Tensor):
        loss = torch.tensor(
            0., device=fake[0][0].device, dtype=fake[0][0].dtype)
        for scale in fake:
            loss += masked_mean_from_ratios(F.relu(1 - scale[-1]), mask_ratio)
        return loss
```
My question is, is there any specific reason you decided to change this loss?

**Comment by yuval-reshef** [comment]
From what I understand, section 2.3.2 of the STREAMVC paper cites [MelGan](https://arxiv.org/pdf/1910.06711) as the base for the the adversarial loss, which uses this formulation (in section 2.3).

**Comment by arishov1** [comment]
Oh, I see. But if you use the formulation from that source, does that mean the discriminator loss is incorrect?

In the formulation, it uses **min⁡(0,1−D_k(x))**, but in the implementation, we use **F.relu(1 - scale[-1])**, which is equivalent to **max⁡(0,1−scale[−1])**.

![image](https://github.com/user-attachments/assets/0bf6c1e1-5e0e-49e8-b8a3-f6788cea6632)

**Comment by yuval-reshef** [comment]
Our implementation is based on [MelGan's official implementation](https://github.com/descriptinc/melgan-neurips/blob/master/scripts/train.py#L143), you can see similar doubts expressed in an [issue](https://github.com/descriptinc/melgan-neurips/issues/34) in the melgan repo, however as per the discussion there, the error is in the paper.

Let's follow the trail of citations to dig a little deeper.

STREAMVC actually cites two papers as the source for the adversarial/feature/reconstruction losses, [MelGan](https://proceedings.neurips.cc/paper_files/paper/2019/file/6804c9bca0a615bdb9374d00a9fcba59-Paper.pdf) and [Streaming-SEANet](https://arxiv.org/pdf/2010.10677).

MelGan's paper shows the adversarial losses as:
$\min\limits_{D_k}  E_x \left[ \min (0, 1-D_k(x)) \right] + E_{s,z} \left[ \min (0, 1+D_k(G(s,z))) \right] ,  \forall k=1,2,3 $
$\min\limits_G E_{s,z}\left[ \sum\limits_{k=1,2,3} - D_k(G(s,z)) \right] $
However it's cites 2 papers: [Geometric GAN](https://arxiv.org/pdf/1705.02894) (which the GAN hinge loss is usually attributed to) and [Spectral Normalization](https://arxiv.org/pdf/1802.05957).

Geometric GAN has the following loss:
$\min\limits_{D}  E_x \left[ \max (0, 1-D(x)) \right] + E_{z} \left[ \max (0, 1+D(G(z))) \right] $
$\min\limits_G -E_{z}\left[  D(G(z)) \right] $
which is similar to our implementation.

The Spectral Normalization paper has the following loss:
$\max\limits_{D}  E_x \left[ \min (0, -1+D(x)) \right] + E_{z} \left[ \min (0, -1-D(G(z))) \right] $
$\min\limits_G -E_{z}\left[  D(G(z)) \right] $
which is exactly the same as the Geometric GAN formulation (which is also cited as inspiration).

Finally, Streaming-SEANet used a similar loss to [SoundStream](https://arxiv.org/pdf/2107.03312):
$\min\limits_{D}  E_x \left[ \max (0, 1-D(x)) \right] + E_{z} \left[ \max (0, 1+D(G(z))) \right] $
$\min\limits_G E_{z}\left[ \max(0, 1- D(G(z))) \right] $
Streaming-SEANet cites [SEANet](https://arxiv.org/pdf/2009.02095) as inspiration, and both SEANet and SoundStream don't cite anyone for the adversarial loss, just call it hinge loss.

So I hope it makes sense why we chose this formulation.

**Comment by arishov1** [comment]
Okay, got it. Thank you for your detailed answer!
